proxy/context.go

```diff
@@ -0,0 +1,178 @@
+package proxy
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/zalando/skipper/routing"
+)
+
+type context struct {
+	responseWriter        http.ResponseWriter
+	request               *http.Request
+	response              *http.Response
+	route                 *routing.Route
+	deprecatedServed      bool
+	servedWithResponse    bool // to support the deprecated way independently
+	pathParams            map[string]string
+	stateBag              map[string]interface{}
+	originalRequest       *http.Request
+	originalResponse      *http.Response
+	outgoingHost          string
+	debugFilterPanics     []interface{}
+	outgoingDebugRequest  *http.Request
+	incomingDebugResponse *http.Response
+	startServe            time.Time
+}
+
+func defaultBody() io.ReadCloser {
+	return ioutil.NopCloser(&bytes.Buffer{})
+}
+
+func defaultResponse(r *http.Request) *http.Response {
+	return &http.Response{
+		StatusCode: http.StatusNotFound,
+		Header:     make(http.Header),
+		Body:       defaultBody(),
+		Request:    r,
+	}
+}
+
+func cloneURL(u *url.URL) *url.URL {
+	uc := *u
+	return &uc
+}
+
+func cloneRequestMetadata(r *http.Request) *http.Request {
+	return &http.Request{
+		Method:           r.Method,
+		URL:              cloneURL(r.URL),
+		Proto:            r.Proto,
+		ProtoMajor:       r.ProtoMajor,
+		ProtoMinor:       r.ProtoMinor,
+		Header:           cloneHeader(r.Header),
+		Body:             defaultBody(),
+		ContentLength:    r.ContentLength,
+		TransferEncoding: r.TransferEncoding,
+		Close:            r.Close,
+		Host:             r.Host,
+		RemoteAddr:       r.RemoteAddr,
+		RequestURI:       r.RequestURI,
+		TLS:              r.TLS,
```

Comment:
Should this also clone the `Form` fields and `Trailer`? 

Also, copying the `ContentLength` but using an empty/default body seems dangerous, as it leaves the request in a potentially invalid state.